src/serlio/modifiers/PRTModifierAction.cpp

```diff
@@ -214,46 +278,248 @@ MStatus PRTModifierAction::fillAttributesFromNode(const MObject& node) {
+
+		switch (attrType) {
+			case PrtAttributeType::BOOL: {
+				const auto defBoolVal = defaultAttributeValues->getBool(fqAttrName.c_str());
+				bool boolVal;
+				MCHECK(plug.getValue(boolVal));
+
+				isDefaultValue = defBoolVal == boolVal;
+				break;
+			}
+			case PrtAttributeType::FLOAT: {
+				const auto defDoubleVal = defaultAttributeValues->getFloat(fqAttrName.c_str());
+				double doubleVal;
+				MCHECK(plug.getValue(doubleVal));
+
+				isDefaultValue = defDoubleVal == doubleVal;
+				break;
+			}
+			case PrtAttributeType::COLOR: {
+				const wchar_t* defColStr = defaultAttributeValues->getString(fqAttrName.c_str());
+
+				MObject rgb;
+				MCHECK(plug.getValue(rgb));
+				MFnNumericData fRGB(rgb);
+
+				prtu::Color col;
+				MCHECK(fRGB.getData3Float(col[0], col[1], col[2]));
+				const std::wstring colStr = prtu::getColorString(col);
+
+				isDefaultValue = std::wcscmp(colStr.c_str(), defColStr) == 0;
+				break;
+			}
+			case PrtAttributeType::STRING: {
+				const auto defStringVal = defaultAttributeValues->getString(fqAttrName.c_str());
+
+				MString stringVal;
+				MCHECK(plug.getValue(stringVal));
 
-			const auto defVal = defaultAttributeValues->getString(fqAttrName.c_str());
-			if (std::wcscmp(val.asWChar(), defVal) != 0)
-				aBuilder->setString(fqAttrName.c_str(), val.asWChar());
+				isDefaultValue = std::wcscmp(stringVal.asWChar(), defStringVal) == 0;
+				break;
+			}
+			case PrtAttributeType::ENUM: {
+				MFnEnumAttribute eAttr(fnAttribute.object());
+
+				short defEnumVal;
+				short enumVal;
+				MCHECK(eAttr.getDefault(defEnumVal));
+				MCHECK(plug.getValue(enumVal));
+
+				isDefaultValue = defEnumVal == enumVal;
+				break;
+			}
+
+			default:
+				break;
 		}
-		else if (attrObj.hasFn(MFn::kEnumAttribute)) {
-			MFnEnumAttribute eAttr(attrObj);
-
-			short di;
-			short i;
-			MCHECK(eAttr.getDefault(di));
-			MCHECK(plug.getValue(i));
-			if (i != di) {
-				switch (ruleAttrType) {
-					case prt::AAT_STR:
-						aBuilder->setString(fqAttrName.c_str(), eAttr.fieldName(i).asWChar());
-						break;
-					case prt::AAT_FLOAT:
-						aBuilder->setFloat(fqAttrName.c_str(), eAttr.fieldName(i).asDouble());
-						break;
-					case prt::AAT_BOOL:
-						aBuilder->setBool(fqAttrName.c_str(), eAttr.fieldName(i).asInt() != 0);
-						break;
-					default:
-						LOG_ERR << "Cannot handle attribute type " << ruleAttrType << " for attr " << fqAttrName;
+
+		if (getAndResetForceDefault(fnNode, fnAttribute)) {
+			setIsUserSet(fnNode, fnAttribute, false);
+		}
+		else {
+			setIsUserSet(fnNode, fnAttribute, !isDefaultValue);
+		}
+	};
+
+	iterateThroughAttributesAndApply(node, updateUserSetAttribute);
+
+	return MStatus::kSuccess;
+}
+
+MStatus PRTModifierAction::updateUI(const MObject& node) {
+	const auto updateUIFromAttributes = [this](const MFnDependencyNode& fnNode, const MFnAttribute& fnAttribute,
+	                                           const RuleAttribute& ruleAttribute, std::wstring fqAttrName,

```

Comment:
RuleAttribute is used in fillAttributesFromNode and all iterator lambda function need the same signature.
I removed the fqAttrName and only pass the ruleAttribute as an argument, because the attributeName is a member of ruleAttribute. This allows us to reduce the argument list, have the same signature for all functions without having unused/redundant arguments